diffkemp/semdiff/caching.py

```diff
@@ -150,6 +150,44 @@ def compare_vertex_priority(self, other):
                     return True
             return False
 
+        def to_yaml(self):
```

Comment:
I'm not sure if doing conversion to YAML inside ComparisonGraph is a good idea, most of result processing in DiffKemp is done elsewhere (in `function_diff.py` and `diffkemp.py`), the idea being that the `Result` class contains the comparison results. Also, it creates some confusion, since ComparisonGraph's `from_yaml` methods refer to a different YAML format (the one generated by SimpLL).

Wouldn't just extracting the required information from `Result` work? If I understand it correctly, the YAML contains mostly the same information as regular DiffKemp output, hence it should be already (maybe with a few expections) present in `Result`.

Reply:
Thanks for the feedback. 

The idea was to create a serialized formatted output with the result of DiffKemp analysis which could be used for  visualization of the results and maybe for other purposes. I am not sure if the format, which I thought up, is the best way for how to do it. I thought that it would be good to print out *call stacks* (only name of functions) of the results (field `diff` in this YAML) and the *call graph* / `ComparisonGraph` (field `names` in this YAML) to not repeat information about functions (which can appear multiple times in call stacks).

I agree that having `from_yaml` and `to_yaml` (which refer to a different format) in the same class (`ComparisonGraph`) is confusing. I am thinking about moving the extraction (conversion to YAML) to a new class (or maybe to `Result` class).

Or maybe I could do it more differently and change the output format - I will try to explain my thoughs.  I could extract more information from `Result` class for creating call stacks (in YAML field `diff`), where I could add `line`, where the function is called. Then from functions which will appear in callstack I could extract from `ComparisonGraph` information about their deffinitions (both names, number of line where function starts, file) - field `names` in YAML. Then the new field `names` would not have to contain fields `calls` and `types`. The new yaml format, after these edits and after renaming some fields, could look something like this.
```
old-snapshot: /home/lukas/BC/diffkemp/kernel-src/SNAPSHOT-80
new-snapshot: /home/lukas/BC/diffkemp/kernel-src/SNAPSHOT-147
# list of functions which are not equal
results:
- function: __alloc_workqueue_key # name of compared function (for example KABI function)
  diffs: # list of functions in which difference is found
  - function: worker  # name of function which is different
    old-callstack: # callstack in old snapshot 
    - name: init_rescuer
      line: 4094 # line where the function is called
    - # ...
    new-callstack:
    - # similar to old-callstack
    - # ...
  - # ..
- # ...

definitions: # list of information about "definitions"
  process_one_work: # name of function/macro/type
    kind: function
    old: #  information about definition in old snapshot
      name: process_one_work
      file: kernel/workqueue.c
      line: 2046
      end-line: 2193
    # information about definition in new snapshot
    new:
      # ... similar to first
  # ...
```
But I am not sure, if the first version of the format where the field `names` represented `ComparisonGraph` would not be more useful in future for some purposes.

I could also remove the field `definitions` (`names`) completely and move these information to functions in call stack  (field `results`), but then i think lot of information would repeat and also I think it is useful to have definitions separately to know which pair of functions belongs together (if I am not mistaken they can have different names caused for example by renaming of funcion).

Reply:
> I agree that having `from_yaml` and `to_yaml` (which refer to a different format) in the same class (`ComparisonGraph`) is confusing. I am thinking about moving the extraction (conversion to YAML) to a new class (or maybe to `Result` class).

Agreed, moving it to `Results` sounds reasonable to me.

> [...]
> But I am not sure, if the first version of the format where the field `names` represented `ComparisonGraph` would not be more useful in future for some purposes.

I kind of like this new format better. It seems more compact and human-readable (though that's not its purpose for sure). There'll be some minor repetition (`line` in callstacks if there are mutliple same callstacks) but I'm ok with that. I'd suggest going with this one, it looks pretty extensible for future needs. Btw. shouldn't callstack entries contain also the file name (so that we don't have to pull that from definitions)?
 
> I could also remove the field `definitions` (`names`) completely and move these information to functions in call stack (field `results`), but then i think lot of information would repeat and also I think it is useful to have definitions separately to know which pair of functions belongs together (if I am not mistaken they can have different names caused for example by renaming of function).

Agreed, let's keep the definitions separated, they may come handy in future (in addition to the advantages that you listed).